### PR TITLE
Verify Request Signature

### PR DESCRIPTION
**Pull request by charles-salmon**
This PR:

* Adjusts the VSCode editor tab size from `4` to `2`.
* Ensures that a status code of 400 is returned if a request does not contain a valid signature.
* Ensures that the `node_modules` folder is ignored during deployment of the cloud function (letting the `@google-cloud/functions-framework` install dependencies).
* Uses Cloud Storage to access the Slack signing secret in encrypted form.
* Uses KMS to decrypt the Slack signing secret.
* Refactors unit tests to reduce fragility due to constructor changes.
* Addresses #2.

### CODE REVIEW

File: src/slack-request-signature-validator.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import crypto from "crypto";
+import { Request } from "express";
+import { inject, injectable } from "inversify";
+import timingSafeCompare from "tsscmp";
+
+import { SecretProvider } from "./secret-provider";
+
+@injectable()
+class SlackRequestSignatureValidator {
```
Comment (kevindrecruit) [inline src/slack-request-signature-validator.ts]: The method `isSignatureValid()` is doing a lot of things so I think it needs to be split up.

The requestSignature and requestTimestamp being set seem to be core requirements so perhaps these can be set & checked in the constructor?

Ideally I would want to see something like this:

```typescript
async function isSignatureValid(): Promise<boolean> {
    if (!this.isTimeStampValid()) { 
        return false;
    }
    const slackSigningSecret = await this.getSlackSigningSecret();
    const hmac = this.createSlackHmac(slackSigningSecret);
    const localSignature = this.createLocalSignature(hmac);

    return timingSafeCompare(this.requestSignature, localSignature);
}
```

File: src/slack-request-signature-validator.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  public async isSignatureValid(): Promise<boolean> {
+    const { headers } = this.request;
+    const requestSignature = headers["x-slack-signature"] as string | undefined;
+    const requestTimestamp = headers["x-slack-request-timestamp"];
+    if (!requestSignature || !requestTimestamp) {
+      return false;
+    }
+
+    const fiveMinutesAgo = Math.floor(Date.now() / 1000) - 60 * 5;
```
[the remark this answers is not in the document]
Comment (charles-salmon) [reply src/slack-request-signature-validator.ts]: Have described these a little bit more clearly, and extracted these details to a `private` method:

```typescript
  private isRecentRequest(requestTimestamp: string): boolean {
    const now = Math.floor(Date.now() / 1000);
    const fiveMinutes = 5 * 60;

    return now - +requestTimestamp < fiveMinutes;
  }
```

File: src/slack-request-signature-validator.ts (hunk 1 above)
Comment (charles-salmon) [reply src/slack-request-signature-validator.ts]: You raise quite a good point. An instance of a `SlackRequestSignatureValidator` does not need to cache an the entire `Request`. Have refactored to:

```typescript
  public async isSignatureValid(): Promise<boolean> {
    if (!this.requestSignature || !this.requestTimestamp) {
      return false;
    }

    if (!this.isRecentRequest()) {
      return false;
    }

    if (
      !timingSafeCompare(this.requestSignature, await this.hashRawRequestBody())
    ) {
      return false;
    }

    return true;
  }
```

File: src/calculate-request-handler.ts [hunk 3]
```diff
@@ -29,16 +36,38 @@ interface TextObject {
...
+
+    if (
+      !REQUIRED_ENVIRONMENT_VARIABLES.every(variable =>
+        Object.keys(process.env).some(key => key === variable)
+      )
+    ) {
+      this.response
+        .status(500)
+        .send(ERROR.REQUIRED_ENVIRONMENT_VARIABLES_NOT_SET);
+      throw new Error(ERROR.REQUIRED_ENVIRONMENT_VARIABLES_NOT_SET);
```
[the remark this answers is not in the document]
Comment (charles-salmon) [reply src/calculate-request-handler.ts]: Sending the response notifies the client of the error. Throwing the error notifies the server.

File: src/calculate-request-handler.ts [hunk 4]
```diff
@@ -29,16 +36,38 @@ interface TextObject {
...
   }
 
-  public handleRequest(): void {
+  public async handleRequest(): Promise<void> {
+    dotenv.config();
+
+    if (
+      !REQUIRED_ENVIRONMENT_VARIABLES.every(variable =>
+        Object.keys(process.env).some(key => key === variable)
+      )
```
[the remark this answers is not in the document]
Comment (charles-salmon) [reply src/calculate-request-handler.ts]: Have extracted to another method.

File: src/calculate-request-handler.ts [hunk 5]
```diff
@@ -29,16 +36,38 @@ interface TextObject {
...
+    ) {
+      this.response
+        .status(500)
+        .send(ERROR.REQUIRED_ENVIRONMENT_VARIABLES_NOT_SET);
+      throw new Error(ERROR.REQUIRED_ENVIRONMENT_VARIABLES_NOT_SET);
+    }
+
+    if (!(await this.slackRequestSignatureValidator.isSignatureValid())) {
```
[the remark this answers is not in the document]
Comment (charles-salmon) [reply src/calculate-request-handler.ts]: Have introduced a local variable to improve readability.

File: src/calculate-request-handler.ts [hunk 6]
```diff
@@ -117,4 +146,4 @@ class CalculateRequestHandler {
   }
 }
 
-export { CalculateRequestHandler };
+export { CalculateRequestHandler, ResponseBody, TextObject };
```
[the remark this answers is not in the document]
Comment (charles-salmon) [reply src/calculate-request-handler.ts]: Have split up the responsibilities of this method more clearly.

File: src/secret-provider.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+  public async decryptSecretFromFile(
+    options: DecryptSecretOptions
+  ): Promise<string> {
+    const { bucketName, filename, cryptoKeyPath } = options;
+
+    const ciphertext = (await this.storageClient
+      .bucket(bucketName)
+      .file(filename)
+      .download())[0].toString("base64");
```
[the remark this answers is not in the document]
Comment (charles-salmon) [reply src/secret-provider.ts]: Have split up the responsibilities of this method more clearly.

File: src/secret-provider.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+    return (await this.kmsClient.decrypt({
+      ciphertext,
+      name: cryptoKeyPath
+    }))[0].plaintext
+      .toString("utf8")
+      .trim();
```
[the remark this answers is not in the document]
Comment (charles-salmon) [reply src/secret-provider.ts]: Have split up the responsibilities of this method more clearly.

File: src/calculate-request-handler.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (charles-salmon) [reply src/calculate-request-handler.ts]: I've refactored this a little. The client doesn't actually need to know that environment variables are not set. So I've instead used the `sendStatus` method which will just return a response containing `Internal Server Error`.

File: src/secret-provider.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    options: DecryptSecretOptions
+  ): Promise<string> {
+    const { bucketName, filename, cryptoKeyPath } = options;
+
+    const ciphertext = (await this.storageClient
+      .bucket(bucketName)
+      .file(filename)
+      .download())[0].toString("base64");
+
+    return (await this.kmsClient.decrypt({
```
[the remark this answers is not in the document]
Comment (charles-salmon) [reply src/secret-provider.ts]: I've updated the `CalculateRequestHandler` to send a response with a status code of `500` in the event of an error.

**Pull request merged**